### ISSUE TITLE
non-(X)HTML extension => file copy to TMP folder ... this breaks relative URLs to CSS, images, etc.

### ISSUE THREAD

**Issue by danielweck** [daisy/ace#243]
This works for unit-tests, because the HTML markup does not refer to satellite resources (which will not be the case for real-world documents):

https://github.com/daisy/ace/blob/be5d8a4154a2a38a07b28475f9e0538d91a0bab3/packages/ace-core/src/checker/checker-chromium.js#L33-L42

...therefore, this seems like a broken/useless logic, right?

Unit tests:

https://github.com/daisy/ace/blob/be5d8a4154a2a38a07b28475f9e0538d91a0bab3/tests/__tests__/regression.test.js#L71-L74

https://github.com/daisy/ace/blob/master/tests/data/issue-122/EPUB/content_001.xml
https://github.com/daisy/ace/blob/master/tests/data/issue-122/EPUB/content_002.ace

**Comment by rdeltour** [comment]
> This works for unit-tests, because the HTML markup does not refer to satellite resources (which will not be the case for real-world documents)

It might not be as bad as you think, as these files are only those referenced from the spine.
Granted, the logic is flawed with SVG spine items.
We should ideally check the file type when parsing the Package Document.
And maybe there's an API in Puppeteer now to tell indicate with which type to serve the file…

**Comment by danielweck** [comment]
Sorry Romain, I don't think I made my point clear: the resources linked from an HTML file that has been renamed+copied to a temporary folder on the filesystem are consequently inaccessible (broken links).
Therefore, the Puppeteer Axe runner will fail to fetch those satellite dependencies, but this may in fact remain undetected if there are no fatal errors (the accessibility evaluation just takes place without CSS stylesheets, images, etc. in the original document).
As the Electron Axe runner depends on an HTTP server, things are a little more hairy in this case (a new static hosting route must be created to the path of the temporary folder inside which the HTML file is renamed).
Bref, renaming+copying such HTML files with erroneous extensions works in unit-tests, but this approach seems unrealistic for real-world use-cases, where documents have linked dependent resources.
Unless I am missing something? Thanks! :)

**Comment by rdeltour** [comment]
OK, got it. Yeah I read too fast and hadn't realized the file was copied _alone_ to a brand new directory. This hack/fix is indeed inherently flawed, and would break with checks depending on CSS, or JS, or iframes, etc.

I guess we didn't have any issues so far since the case would be pretty rare (EPUBCheck catches that upfront). But yes I agree this should be fixed. Ideally we would serve the file to Puppeteer as HTML without renaming it.

**Comment by danielweck** [comment]
Right, with the Electron-based Axe runner we can use the HTTP server to enforce correct content-types, but we would have to double-check that the lack of `*.xhtml` extension doesn't trip-up Chromium (this used to be a problem with the polyglot XHTML thing in EPUB ... hopefully the `application/xhtml+xml` trick will be sufficient).

I don't know about Puppeteer.

**Comment by rdeltour** [comment]
> I don't know about Puppeteer.

Right now we merely open the page using the `Page.goto()` API, which doesn't provide a way to override the HTTP headers.

I guess we would have to serve the page with an HTTP server (like we do for the Electron runner) to have more control. It would also probably be closer to how most RS serve the EPUB content docs…

**Comment by danielweck** [comment]
What about something like:

```javascript
await page.setRequestInterception(true);
page.on("request", (req) => {
  const url = req.url();
  if (SOME_URL_TEST(url)) {
    req.respond({
      headers: { ... }
      body: fs.readFileSync(
        path.join(__dirname, TMP_FOLDER, url.replace("file://", ""))
      )
    });
  } else {
    req.continue();
  }
});
```

**Comment by rdeltour** [comment]
yes, could work!
